### PR TITLE
detached coprocessor execution

### PR DESCRIPTION
**Pull request by Geal**
Fix https://github.com/apollographql/router/issues/3297

This adds a new `detached` option at all coprocessor stages to allow the router to continue handling the request or response without waiting for the coprocessor to respond. This implies that the coprocessor response will not be used. This is targeted at use cases like logging and auditing, which do not need to block the router

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: .changesets/feat_geal_async_coprocessor.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+### asynchronous coprocessor execution ([Issue #3297](https://github.com/apollographql/router/issues/3297))
+
+This adds a new `asynchronous` option at all coprocessor stages to allow the router to continue handling the request or response without waiting for the coprocessor to respond. This implies that the coprocessor response will not be used. This is targeted at use cases like logging and auditing, which do not need to block the router
```
Comment (bnjjj) [inline .changesets/feat_geal_async_coprocessor.md:3]: Could you add a full example of configuration with this new option please ?

File: apollo-router/src/plugins/coprocessor/mod.rs [hunk 2]
```diff
@@ -778,6 +806,84 @@ where
...
+                    );
+                });
+            }
+
+            deferred_response
+        });
+
+        // Create our response stream which consists of the bytes from our first body chained with the
+        // rest of the responses in our mapped stream.
+        //let bytes = hyper::body::to_bytes(body).await.map_err(BoxError::from);
```
Comment (bnjjj) [inline apollo-router/src/plugins/coprocessor/mod.rs:880]: to delete ?

File: docs/shared/coproc-typical-config.mdx [hunk 3]
```diff
@@ -4,26 +4,46 @@ coprocessor:
...
     request: # By including this key, the `RouterService` sends a coprocessor request whenever it first receives a client request.
+      asynchronous: false # if set to true, the router will handle the request without waiting for the coprocessor to respond
       headers: true # These boolean properties indicate which request data to include in the coprocessor request. All are optional and false by default.
       body: false
       context: false
       sdl: false
       path: false
       method: false
     response: # By including this key, the `RouterService` sends a coprocessor request whenever it's about to send response data to a client (including incremental data via @defer).
+      asynchronous: false # if set to true, the router will handle the response without waiting for the coprocessor to respond
```
Comment (bnjjj) [inline docs/shared/coproc-typical-config.mdx:15]: Maybe you can document it's optional in the comment

Comment (Geal) [reply docs/shared/coproc-typical-config.mdx]: fixed in [db39565](https://github.com/apollographql/router/pull/4902/commits/db395654b60678a5348d065b56615d86c9c0e41c)

File: apollo-router/src/plugins/coprocessor/mod.rs (hunk 2 above)
Comment (Geal) [reply apollo-router/src/plugins/coprocessor/mod.rs]: fixed in [db39565](https://github.com/apollographql/router/pull/4902/commits/db395654b60678a5348d065b56615d86c9c0e41c)

File: .changesets/feat_geal_async_coprocessor.md (hunk 1 above)
Comment (Geal) [reply .changesets/feat_geal_async_coprocessor.md]: fixed in [db39565](https://github.com/apollographql/router/pull/4902/commits/db395654b60678a5348d065b56615d86c9c0e41c)

File: docs/shared/coproc-typical-config.mdx [hunk 4]
```diff
@@ -32,13 +52,15 @@ coprocessor:
   subgraph:
     all:
       request: # By including this key, the `SubgraphService` sends a coprocessor request whenever it is about to make a request to a subgraph.
+        asynchronous: false # optional. if set to true, the router will handle the request without waiting for the coprocessor to respond
```
Comment (o0Ignition0o) [inline docs/shared/coproc-typical-config.mdx:55]: the router will proceed ?

Does asynchronous mean `fire and forget`?

**Review by bnjjj — APPROVED** [review]

**Review by o0Ignition0o — APPROVED** [review]

**Review by garypen — COMMENTED** [review]
For some of the reasons I documented in the issue, I don't think this is a good approach to solving this problem.

I'd really prefer us to think about how to handle this from an architectural standpoint rather than shoe-horning behaviour into what is, essentially, an RPC mechanism which expects a response and which is co-processor specific.

One way we could address this kind of requirement is by supporting an externalizable event mechanism to which plugins (or any external program, including coprocessors) could subscribe.

If this is really the way that we think we should implement an external event system, then let's not call it "detached", but maybe call it "event"?

One thing I don't like is that the implementor of a coprocessor doesn't know what the router will do with any response that they provide. There's no indication in the payload that an invocation is "detached" or not. Since the behaviour would only be controlled by router configuration, that seems like a potential footgun in the making.

Another thing, which I allude to above, is that if this is a good idea for coprocessor implementations, then isn't it also a good idea to be able to process events from `Rhai` scripts or indeed in Rust plugins?

Summary: I'd rather we implemented a generalised event mechanism.

**Comment by Geal** [comment]
@garypen sorry, Idid not see your message until after this was merged. While a mechanism specifically focused on events could be interesting, I'd worry about introducing yet another way to extend the router and communicate with other services. Especially since we are currently shipping customizable events in telemetry.
I think you're right though that detached coprocessor requests should have a field to indicate that it does not expect an answer

**Pull request merged**